core/src/smartcontracts/isi/permissions.rs

```diff
@@ -20,65 +23,362 @@ impl NeedsPermission for QueryBox {}
 // Expression might contain a query, therefore needs to be checked.
 impl NeedsPermission for Expression {}
 
-/// Reason for prohibiting the execution of the particular instruction.
-pub type DenialReason = String;
+/// Type of object validator can check
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum ValidatorType {
+    /// [`Instruction`] variant
+    Instruction,
+    /// [`QueryBox`] variant
+    Query,
+    /// [`Expression`] variant
+    Expression,
+}
+
+impl ValidatorType {
+    /// Checks if `self` equals to `another`
+    ///
+    /// # Errors
+    /// If `self` doesn't equal to `another`
+    pub fn check_equal(
+        self,
+        another: ValidatorType,
+    ) -> std::result::Result<(), ValidatorTypeMismatch> {
+        if self != another {
+            return Err(ValidatorTypeMismatch::expected(another).found(self));
+        }
+
+        Ok(())
+    }
+}
+
+impl std::fmt::Display for ValidatorType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ValidatorType::Instruction => write!(f, "Instruction"),
+            ValidatorType::Query => write!(f, "Query"),
+            ValidatorType::Expression => write!(f, "Expression"),
+        }
+    }
+}
+
+pub mod error {
+    //! Contains errors structures
+
+    use super::ValidatorType;
+
+    /// Reason for prohibiting the execution of the particular instruction.
+    #[derive(Debug, Clone, thiserror::Error)]
+    #[allow(variant_size_differences)]
+    pub enum DenialReason {
+        /// [`ValidatorTypeMismatch`] variant
+        #[error("{0}")]
+        ValidatorTypeMismatch(#[from] ValidatorTypeMismatch),
+        /// Variant for custom error
+        #[error("{0}")]
+        Custom(String),
+        /// Variant used when at least one [`Validator`](super::IsAllowed) should be provided
+        #[error("No validators provided")]
+        NoValidatorsProvided,
+    }
+
+    impl From<String> for DenialReason {
+        fn from(s: String) -> Self {
+            Self::Custom(s)
+        }
+    }
+
+    /// Wrong validator expectation error
+    ///
+    /// I.e. used when user tries to validate [`QueryBox`](super::QueryBox) with
+    /// [`IsAllowedBoxed`](super::IsAllowedBoxed) containing
+    /// [`IsAllowedBoxed::Instruction`](super::IsAllowedBoxed::Instruction) variant
+    #[derive(Debug, Copy, Clone, thiserror::Error)]
+    #[error("Expected `{expected}` validator type, but found `{found}`")]
+    pub struct ValidatorTypeMismatch {
```

Comment:
We also have a generic `Mismatch<T>` in the `data_model`. We could add a few trait bounds and perhaps make it produce the kind of output that you want, using e.g. `std::any::type`'s string representation.

Reply:
I think `Mismatch<T>` should be fine just with additional info in `DenialReason`